sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java

```diff
@@ -14,82 +17,189 @@
+         *
+         * @param storageBlobClient the storage blob client
+         * @return Builder with provided Storage Blob Client set
+         */
+        public Builder setStorageClient(@NonNull StorageBlobClient storageBlobClient) {
+            this.storageBlobClient = storageBlobClient;
+            return this;
+        }
+
+        /**
+         * Sets the particular {@link TransferNetworkType} the device should be in
+         * for running the transfers.
+         *
+         * @param networkType The type of network required for transfers to run
+         * @return Builder with provided network type set
+         */
+        public Builder setRequiredNetworkType(@TransferNetworkType int networkType) {
+            this.networkType = networkType;
+            return this;
+        }
+
+        /**
+         * Set the {@link Executor} used by {@link TransferClient} for all its internal
+         * book keeping, which includes creating DB entries for transfer workers, querying DB
+         * for status, submitting transfer request to {@link WorkManager}.
+         *
+         * TransferClient will enqueue maximum of two command to the serialBackgroundExecutor at any time.
+         *
+         * @param executor the serialBackgroundExecutor
+         * @return Builder with provided serialBackgroundExecutor set
+         */
+        public Builder setBookKeepingExecutor(@NonNull Executor executor) {
+            this.serialBackgroundExecutor = new SerialExecutor(executor);
+            return this;
+        }
+
+        /**
+         * @return A {@link TransferClient} configured with settings applied through this builder
+         */
+        public TransferClient build() {
+            if (this.storageBlobClient == null) {
+                throw new IllegalArgumentException("storageBlobClient must be set.");
+            }
+            Constraints.Builder constraintsBuilder = new Constraints.Builder();
+            switch (this.networkType) {
+                case TransferNetworkType.ANY:
+                    constraintsBuilder.setRequiredNetworkType(NetworkType.CONNECTED);
+                    break;
+                case TransferNetworkType.UNMETERED:
+                    constraintsBuilder.setRequiredNetworkType(NetworkType.UNMETERED);
+                    break;
+                case TransferNetworkType.METERED:
+                    constraintsBuilder.setRequiredNetworkType(NetworkType.METERED);
+                    break;
+                case TransferNetworkType.NOT_ROAMING:
+                    constraintsBuilder.setRequiredNetworkType(NetworkType.NOT_ROAMING);
+                    break;
+                default:
+                    throw new IllegalArgumentException("networkType must be set.");
+            }
+            if (this.serialBackgroundExecutor == null) {
+                this.serialBackgroundExecutor =  new SerialExecutor(Executors.newFixedThreadPool(2));
```

Comment:
How would you feel about using WorkManager's `Configuration.getTaskExecutor()` here as the default? The docs say that this function "Gets the Executor used by WorkManager for all its internal business logic".

Reply:
(We would still need to wrap it in a `SerialExecutor`)

Reply:
that sounds better, let me check task manager is accessible in app.

Reply:
I don't see there is a way to access the current configuration of WorkManager, though it is settable during application [bootstrap](https://developer.android.com/topic/libraries/architecture/workmanager/advanced/custom-configuration#implement-configuration-provider)

Reply:
Yeah, once again I see plenty of ways to do it but only if we cask to `WorkManagerImpl`. It's so frustrating that so many improvements could be made if just a few things were not hidden from the public API.

Reply:
true, I felt the same for couple of core types.

Reply:
Maybe it's best we just run on our own thread. But I'm wondering - if most of the bookkeeping is just interacting with Room, and room uses its own thread pool, do we actually need to have a task executor? It's not a rhetorical question - I honestly don't know. Is there much more to the bookkeeping than that?

Reply:
The upload method is called from UI thread, even though room may have it's own pool internally, still we're not allowed to make room queries from UI thread, room api throws if it find that query is initialed from main thread.

This means, we need to offload the work to a different thread. Question is,  should we spin up a new thread per transfer API call for bookkeeping vs should we hold a thread pool to do these bookkeeping.

Spinning and tearing down short lived threads per transfer api call could add overhead, given thread creation is not cheap.

Having "our own thread pool" can save thread creation cost, but it still hold threads that is only used for bookkeeping purposes and not accessible for customers.

Having a delegated TaskExecutor, enables us to still do the bookkeeping, at the same time we leave an option for customers to share their executor with us that they also use for some their own work. We use it carefully, in the sense, we don't overwhelm it with too many works in parallel given there is a serialization.

There seems many trade-off here across options.

Reply:
Seems so. I guess it's fine if we have our own small task executor, and if a customer wants to optimize they can create their own executor and pass it to both the configuration of WorkManager / Room and to our TransferClient.

I still feel like the optimal solution is to get the current task executor from WorkManager (whether the default or the user-supplied one) and reuse it. There's just no way to do that with only their public API. For everything else I am good with things being sub-optimal if it means sticking with only the public API (e.g. duplicating the SerialExecutor class). However in this one case I do wonder if the tradeoff is worth it. Having an extra thread pool just for our own book keeping is certainly an overhead.

Reply:
agree, I really wish we have a stable way to  retrieve `WorkManager` executor, which is optimal but so bad that they hide it.

Just double checking, so do you agree with current sub-optimal soln?

1. Internally we use `SerialExecutor `, we don't leak `SerialExecutor ` type through TransferClient public APIs.
2. We allow user to share their `Executor` with us, to which we serially delegate the bookkeeping work.
3. If user don't provide one then we internally create `Executor` and use for book keeping.

Reply:
#1 yes, I'm fine with us having our own copy of SerialExecutor to avoid having to use non-public API

#2-3, what do you think of this proposal for `TransferClient.build()`:

```java
if (this.serialBackgroundExecutor == null) {
    try {
        // Try to re-use the existing taskExecutor shared by WorkManager and Room
        WorkManagerImpl wmImpl = (WorkManagerImpl)WorkManager.getInstance(context);
        this.serialBackgroundExecutor = new SerialExecutor(wmImpl.getConfiguration().getTaskExecutor());
    } catch (Throwable ignored) {
        // Create our own small ThreadPoolExecutor if we can't
        this.serialBackgroundExecutor = new SerialExecutor(Executors.newFixedThreadPool(2));
    }
}
```

In the above, we're unsafely using the non-public API of WorkManager, but if it fails we catch any type of Throwable and create our own thread pool instead.

Reply:
cool, I'm fine with doing this best effort to avoid creating our own pool.

So if ever WM remove this accessor then as part of upgrading our libs to the new WM, we can adjust our code.

If customer override WM dependency in their application, such that the version they bring is different from the one we use and if their WM is missing this accessor, then it will be MethodNotFoundException which we catch here (as RuntimeException) and fall back to our own thread pool.

Reply:
I could also imagine a future change causing ClassCastException or NullPointerException as well but all of those should be caught here. Just documenting for posterity :)